[REFACTOR] 회원가입 폼 리팩터링을 위한 테스트 추가 및 상태관리 전략 개선

kimyou1102 opened the pull request:
## Issue Number
closed #733

## As-Is
<!-- 문제 상황 정의 -->
- 리팩토링시 버그 발생 체크를 위한 테스트 부재
- 인풋값의 변화로 인한 모든 에러메시지를 파생상태로 관리하여 UI변경 로직의 복잡성 증가
- 폼 엔터시 올바른 인증번호 입력해도 에러메시지 뜨던 버그
- 인증 후 새로운 번호로 변경시 제출하기 버튼 클릭시 에러메시지가 뜨지 않던 버그

## To-Be
<!-- 변경 사항 -->
- 회원가입 폼 테스트코드 추가
- 상태관리 전략 파생상태 + 원본상태 로 변경
- 2건의 버그 픽스

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
### 상태관리 전략 변경: 파생상태 -> 원본상태 + 파생상태 전략

이번 PR에서는 **회원가입 폼 리팩터링을 위한 사전 작업**으로, 테스트 추가 과정에서 확인된 문제를 계기로 **에러 및 검증 상태 관리 방식을 명확히 분리**했습니다

### 공통 개선 방향
- 기존에 `useSubmitGuardWithConfirm` 기반으로 얽혀 있던 **복합 파생 상태 구조를 제거**
- 검증 로직을 다음 기준으로 재정의
    - **입력값으로부터 계산 가능한 에러메시지 → 파생 상태**
    - **사용자 행동(버튼 클릭, 검증 시도 여부 등)으로만 알 수 있는 값 → 원본 상태**
- 최종적으로 화면에 노출되는 에러 메시지는 **원본 상태 + 파생 상태를 조합한 함수**에서 결정하도록 정리
- 버튼 클릭, 입력 변경 등 **사용자 행동에 따른 상태 전이가 명확해지고** 테스트 작성 시에도 “어떤 상태가 어떤 메시지를 만드는지” 추적하기 쉬워졌습니다.

### 아이디 중복 확인 검증 로직 개선
- 사용자 인터렉션을 통한 값을 명확히 표현하기 위한 메타 상태 도입
    - `userIdDuplicateChecked`: 중복 확인 완료 여부
    - `submitAttempted`: 제출 시도 여부
- `handleSubmit`에서 `submitAttempted`를 기준으로 중복 확인 없이 제출되는 흐름을 명확히 차단하도록 개선
- 최종 에러 메시지는 **입력 예외 → 중복 실패 → 미시도 상태** 순으로 우선순위를 두고 `getFinalUserIdErrorMessage`에서 계산하도록 수정했습니다.
```tsx
const getFinalUserIdErrorMessage = () => {
  if (userIdErrorMessage !== '') {
    return userIdErrorMessage;
  }

  if (duplicateError) {
    return '이미 사용 중인 아이디입니다.';
  }

  if (!userIdDuplicateChecked && submitAttempted) {
    return '중복확인을 해주세요';
  }

  return '';
};
```

### 전화번호 인증 / 인증번호 확인 로직 개선
- 인증 단계를 `verificationStep` 상태(`idle → requested → verified`)로 명확히 정의
- 인증번호 입력 변경 시 verificationCodeError 상태를 초기화하여 재입력 시 에러 상태가 자연스럽게 해제되도록 처리
- 최종 전화번호 에러 메시지는 제출 시도 여부 및 인증 단계 → 입력 예외 검사 순으로 우선순위를 두고  getFinalPhoneNumberErrorMessage에서 단일 기준으로 계산하도록 정리
```tsx
  const getFinalPhoneNumberErrorMessage = () => {
    if (submitAttempted && verificationStep === 'idle') {
      return '인증요청을 해주세요.';
    }

    return phoneNumberErrorMessage;
  };
```
- 최종 인증번호 에러 메시지는 **입력 예외 → 서버 검증 실패 → 미검증 상태(요청만 된 상태)** 순으로 우선순위를 두고 `getFinalVerificationCodeErrorMessage`에서 계산하도록 정리
```tsx
  const getFinalVerificationCodeErrorMessage = () => {
    if (verificationCodeErrorMessage) {
      return verificationCodeErrorMessage;
    }

    if (verificationStep === 'verified' && verificationCodeError) {
      return '인증 실패';
    }

    if (submitAttempted && verificationStep === 'requested') {
      return '인증을 해주세요';
    }

    return '';
  };
```

###  MSW 핸들러 매칭 이슈 수정

**문제 상황**
테스트 환경에서 MSW 핸들러의 최종 URL에 사용하던 도메인을 환경변수로 불러오는 과정에서 값이 undefined가 되었고, 그로 인해 MSW 핸들러 URL과 실제 API 요청 URL이 매칭되지 않아 MSW 핸들러 대신 **실제 API 요청이 호출되는 문제가 발생**했습니다.

**해결 방법**
도메인 값에 의존하지 않도록 MSW 핸들러의 매칭 방식을 변경해, URL 전체가 아닌 path 기준으로 요청을 매칭하도록 와일드카드를 적용했습니다.
이를 통해 도메인 환경과 무관하게 path만 일치하면 MSW 핸들러가 정상적으로 동작하도록 수정했습니다.

kimyou1102 committed:
refactor: BASE_URL 의존성 제거를 위해 핸들러 URL에 와일드카드 적용 #733.

- 테스트 실행 시, 환경변수가 로드되기 전에 핸들러 모듈이 먼저 평가되어 URL이 'undefined/api/...' 형태로 설정되는 문제 발생
- 핸들러 URL 경로 앞에 와일드카드(*)를 사용하여 BASE_URL 값과 무관하게 요청을 안정적으로 가로챌 수 있도록 수정

kimyou1102 committed:
test: 회원가입폼 입력 검증 테스트 작성 #733

kimyou1102 committed:
test: SignupForm 입력/비밀번호/전화번호 인증 시나리오 테스트 작성 #733

kimyou1102 committed:
fix: dotenv 로딩 경로를 절대 경로로 수정 #733

- 로컬 환경에서 테스트 실행 시, 터미널의 현재 작업 디렉토리(CWD) 위치에 따라 `.env` 파일을 찾지 못해 환경 변수를 로드하지 못하는 문제 발생
- `path.resolve(process.cwd(), ...)`를 사용하여 프로젝트 루트를 기준으로 하는 절대 경로로 변경

kimyou1102 committed:
refactor: userId 중복검증 훅 및 submit 로직 리팩토링 #733

- 이전 구조는 useSubmitGuardWithConfirm 훅을 통한 파생 상태 관리였으나,
    버튼 클릭 시 에러 메시지 출력이 헷갈리고 확장성도 낮음
- useSubmitGuardWithConfirm 기반 파생 상태 구조 제거
- 중복확인 체크를 위한 userIdDuplicateChecked 상태와 제출 시도여부를 위한 submitAttempted 상태 도입
- 입력 예외 검증은 파생 상태, 중복 검증/시도 여부는 원본 상태로 분리
- handleSubmit에서 submitAttempted 기반 중복 체크 후 제출 처리

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/733-signupform-refactor

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/733-signupform-refactor

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/733-signupform-refactor

kimyou1102 committed:
test: 인증요청 클릭하지 않을시 인증하기 버튼 활성화되지 않도록 테스트 검증 변경 #733

kimyou1102 committed:
refactor: 전화번호 인증 및 인증번호 확인 로직 변경 #733

- 불필요한 useSubmitGuardWithConfirm 훅을 사용하지 않게 되어서 handler함수만 훅안에 존재하게됨.
- 훅을 사용하지 않고 handler함수를 SignupForm에서 바로 사용
- verificationStep이 verified가 아닐때 모두 return하는 조건문이 있기 때문에 전화번호 체크, 인증번호 체크 조건문 제거
- 인증번호 입력 인풋 체인지마다 verificationCodeError 상태를 false로 초기화

kimyou1102 committed:
test: 테스트 설명 및 테스트 검정 변경 #733

- 버튼은 활성화 되어있고 에러메시지만 뜨도록 구현해서 이에 맞게 변경

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/733-signupform-refactor

kimyou1102 committed:
refactor: 변경된 폴더구조에 맞게 import문 수정 #733

kimyou1102 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2025-Fit-toring into feature/733-signupform-refactor

kimyou1102 committed:
test: customRender로 변경 #733

kimyou1102 committed:
test: 인증번호 입력 관련 테스트에서 사전 입력 준비 과정 추가 #733

kimyou1102 committed:
fix: 인증하기 버튼 활성화 조건 로직 변경 #733

- requested 단계에서 조건을 검증하지 않아 버튼이 항상 활성화됨
- requested/verified 공통으로 입력 검증을 적용해 올바른 상태에서만 활성화되도록 수정

kimyou1102 committed:
refactor: msw핸들러 url을 와일드 카드로 변경 #733

dlsxjzld commented on frontend/setupTests.ts:
```diff
@@ -1,11 +1,13 @@
 import { expect, afterEach, beforeAll, afterAll } from 'vitest';
 
-config({ path: '.env.prod' });
+config({ path: path.resolve(process.cwd(), '.env.prod') });
```
p5: 혹시 path를 추가하신 이유가 있을까요?? 🤔

kimyou1102 replied:
msw 핸들러와 매칭되지 않던 원인을 해결하는 과정에서 dotenv 경로 문제를 의심해 여러 방법을 시도하며 추가했던 코드입니다. 핸들러에 와일드카드를 적용하면서 해결되어 현재는 불필요한 코드라서 제거하겠습니다!

kimyou1102 replied:
[refactor: dotenv config 경로를 절대에서 상대 경로로 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/4a75c2143f86ac233f61a176b37f81867ec74911)

dlsxjzld commented on frontend/src/common/mock/authCode/handlers.ts:
```diff
@@ -2,8 +2,7 @@ import { http, HttpResponse } from 'msw';
 
-const BASE_URL = process.env.API_BASE_URL;
-const AUTH_CODE_URL = `${BASE_URL}${API_ENDPOINTS.AUTH_CODE}`;
+const AUTH_CODE_URL = `*${API_ENDPOINTS.AUTH_CODE}`;
```
p4: 이 부분은 왜 *를 사용하고 BASE_URL이 없어지게 된건지 궁금합니다 🤔

PR에도 설명이 있으면 더 좋을 것 같아요 ㅎㅎ

kimyou1102 replied:
테스트 환경에서 MSW 핸들러의 최종 URL에 사용하는 도메인을 환경변수로 불러오는 과정에서 값이 undefined가 되었고, 그로 인해 MSW 핸들러와 실제 API 요청 URL이 매칭되지 않아 MSW 핸들러 대신 실제 API 요청이 호출되는 문제가 발생했습니다.

이를 해결하기 위해 도메인과 무관하게 path 기준으로 요청을 매칭하도록 와일드카드를 적용해, path만 일치하면 MSW 핸들러가 동작하도록 수정했습니다!

PR본문에도 와일드카드로 변경한 이유에 대해 추가하도록 하겠습니다🙂

dlsxjzld replied:
와일드카드로 해결하셨군요 👍👍

개인적으로 궁금해서 좀 더 원인을 찾아봤습니다 ㅎㅎ

### 테스트 환경에서 BASE_URL이 undefined가 되었던 이유
- 모듈 시스템의 실행 순서 때문인 것 같습니다!
- setupTests.ts에서 config()를 호출하더라도, 자바스크립트 엔진이 **import 구문을 호이스팅 시켜서 먼저 실행**하기 때문에 handler.ts가 로드되는 시점에는 아직 환경변수가 주입되지 않은 것 같습니다.

### 와일드카드 사용
- 와일드 카드로 path 매칭을 해결할 수 있지만, 다음과 같은 이유로 지양하는 게 좋을 것 같아요.
- MSW 핸들러는 테스트뿐만 아니라 브라우저 로컬 개발(npm run dev) 시에도 사용됩니다. 만약 도메인을 명시하지 않고 path만으로 매칭하면, 우리 서버가 아닌 외부 서비스(예: 카카오 로그인 등)로 보내는 요청까지 MSW가 의도치 않게 가로채버리는 문제가 발생할 수도 있을 것 같아요!
   1. 우리 서버 로그인: https://my-api.com/login
       - MSW가 가로채야 함 -> 성공 (의도대로 동작)

    2. 카카오/구글 로그인 (외부 API): https://kauth.kakao.com/oauth/login ...
        - 경로 끝에 /login이 포함되어 있음.
        - MSW: "어? 경로가 /login이네? 내가 처리할게!" 하고 가로채버림.
        - 결과: 실제 카카오 서버로 요청이 안 날아가고 MSW가 가짜 응답을 줘버리거나 에러가 남. 

### 해결 방법
- Vitest가 테스트 파일을 로드하기 전에 환경변수를 확실하게 주입하도록 vitest.config.ts 설정을 수정하는 방법을 제안합니다.
- setupTests에 있는 config 코드를 vitest.config.ts에 옮기면 될 것 같아요!

```ts
import path from 'path';

import { config } from 'dotenv';
import { defineConfig } from 'vitest/config';
config({ path: path.resolve(process.cwd(), '.env.test') });

export default defineConfig({
  test: {
    globals: true,
    environment: 'jsdom',
    setupFiles: './setupTests.ts',
    css: true,
    env: {
          API_BASE_URL: process.env.API_BASE_URL,
        },
  },
});
```

### 추가 제안
테스트에 대해 고민하다보니 **.env.test를 새로 생성**하는 게 나을 것 같다는 생각이 들었습니다.
- 안전성: 테스트 중에 실수로 MSW가 꺼지더라도, dev나 prod 실제 서버(DB)로 요청이 날아가는 대참사를 막기 위해 가짜 도메인(http://localhost:8080 등)을 사용하는 것이 안전합니다.

- 격리: 로컬이든 CI든 항상 NODE_ENV=test로 잡히므로, 외부 환경에 의존하지 않는 독립적인 테스트가 가능해집니다.

### 참고자료
- [vitest에서 vite 사용하여 env 설정하기](https://vitest.dev/guide/features.html#environment-variables): 현재 vite대신 웹팩을 사용하므로 dotenv를 사용하여 대체하려고 했습니다.

kimyou1102 replied:
와일드카드로 문제를 임시 해결하는것보다 더 근본적인 해결책인거 같아서 좋습니다!👍👍 꼼꼼히 찾아봐주셔서 감사합니다🙂

kimyou1102 replied:
[refactor: MSW 핸들러 URL에서 환경변수를 읽기 위해 env 로딩을 vitest.config.ts로 이동](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/1dcc31a427bf9a0e2e25e7627262d9aeb5a1f1ea)
[refactor: MSW 핸들러 URL 와일드카드 제거](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/aca2dddb2c7a3cf8fb5e97202724674f8b8af15d)

현재는 env.prod가 읽히도록 했고, env.test로 변경하는 것은 pr을 파서 따로 올리겠습니다!

dlsxjzld replied:
좋습니다!!! .env.test로 변경하는 건 제가 [다른 PR](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1157)에서 하겠습니다 👍

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -55,6 +55,9 @@ function SignupForm() {
     setGender(value);
   };
 
+  // 아이디
```
p3: 주석이 잘못 들어간 것 같아요 😊

kimyou1102 replied:
[refactor: 불필요한 주석 제거](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/c8d7015fbbcfd04da0793487b8d499a326f8ecdd)

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -55,6 +55,9 @@ function SignupForm() {
 
+  // 아이디
+  const [userIdDuplicateChecked, setUserIdDuplicateChecked] = useState(false);
+
```
p3: 혹시 useUserIdDuplicateCheck 훅의 duplicateChecked와 어떤 차이점이 있는지 알려주실 수 있을까요?? 🤔

userIdDuplicateChecked과 duplicateChecked이 같은 역할을 하고 있는 것 같아서 궁금합니다.

kimyou1102 replied:
현재는 중복확인 로직이 “중복확인 버튼을 눌러 성공했는가” 기준이라, 아이디 입력 input이 변경되었을 때 기존 확인 결과를 무효화하기 위해 userIdDuplicateChecked를 추가로 두었습니다.

다만 말씀 주신 것처럼 역할이 겹치는 부분이 있어서, 확인된 userId를 기준으로 중복확인 여부를 판단하도록 로직을 정리해 하나의 상태로 관리하는 식으로 변경해보겠습니다!

kimyou1102 replied:
[refactor: 아이디 중복확인 상태를 duplicateChecked상태로 판단하도록 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/71b0c1fc719b035d19ca4d422056b5780a4f24c5)

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -65,20 +68,25 @@ function SignupForm() {
+
     if (duplicateChecked) {
       resetDuplicateCheck();
     }
```
p4: 혹시 duplicateChecked으로 duplicateError를 대신할 수 있는지 궁금합니다! 🤔

API 명세서의 아이디 중복 검증 요청을 봤습니다!
- 중복된 아이디가 아니면 200을 반환: duplicateChecked를 true로 변경
- 중복된 아이디면 400을 반환: duplicateError를 true로 변경

duplicateChecked가 다음과 같은 의미를 나타낸다고 이해했습니다.
- false면 중복된 아이디
- true면 중복되지 않은 아이디

duplicateChecked 만으로도 duplicateError의 역할을 다할 수 있을 것 같은데 어떻게 생각하시나요?
혹시 duplicateError가 필요한 이유가 있을까요? 🤔

kimyou1102 replied:
말씀해주신 것처럼 API 명세 기준으로만 보면 `duplicateChecked` 하나로도 표현할 수 있을 것처럼 보입니다.

다만 현재 로직에서는
- 아직 중복 확인을 하지 않은 상태
- 중복된 아이디인 경우(400)
- 네트워크/서버 에러로 요청이 실패한 경우

이 세 가지가 모두 `duplicateChecked = false`가 되어 UI와 검증 로직에서 원인을 구분하기 어려워집니다.
그래서 요청 실패(에러) 상태를 명확히 표현하기 위해`duplicateError`를 분리해 두었습니다!

dlsxjzld replied:
1. 그렇다면 duplicateError가 true가 되는 경우는 아래 2가지 경우일까요? 🤔
    1. 중복된 아이디인 경우
    2. 네트워크/서버 에러로 요청이 실패한 경우 

2. 아직 중복 확인을 하지 않은 상태는 duplicateChecked만으로 확인을 하는 것 맞을까요? ㅎㅎ

kimyou1102 replied:
넵 맞습니다!

dlsxjzld replied:
넵 감사합니다 ☺️

dlsxjzld commented on frontend/src/common/mock/authCodeVerify/handlers.ts:
p3: 혹시 msw에 인증번호가 실패하는 조건도 추가하는 건 어떻게 생각하시나요?? 😊

msw를 키고 테스트를 했는데 이런 경우를 발견했어요.

<img width="722" height="616" alt="인증번호틀렸는데가입됨" src="https://github.com/user-attachments/assets/450c4eeb-0c0c-4e6d-a276-abd8090deae6" />

1. 123458 입력
2. MSW가 아무 응답도 안 함 
3. response.status === 200 체크 실패 → alert 안 뜸
4. catch로 안 감 (에러가 아니라 그냥 status가 200이 아닐 뿐)
5. finally에서 completeVerification()실행!
6. verificationStep이 'verified'가 됨
7. 제출 가능!

실패하는 조건에 대한 처리가 있으면 탄탄할 것 같습니다!

kimyou1102 replied:
123458을 입력하고 
인증하기를 누르면 아래와 같이 인증 실패가 뜨고
<img width="446" height="282" alt="image" src="https://github.com/user-attachments/assets/7e556f05-b645-4950-b9ac-aeb96656a478" />


인증하기를 누르지않고 회원가입을 누르면 인증을 해주세요라는 메시지가 뜨는데 위와 같은 경우는 어떤 경우였을지 더 자세히 말해주실 수 있을까요?
<img width="456" height="291" alt="image" src="https://github.com/user-attachments/assets/2dd1b851-5d75-4e95-98e6-4e6b940e6066" />

버튼은 막히지만 엔터로 제출은 되는 상황을 발견해서 이에 대해서는 막아두도록 하겠습니다! 꼼꼼한 리뷰 감사합니다👍👍

dlsxjzld replied:
다시 확인하니 잘못된 방법을 올렸었네요..!
시간을 뺏어서 죄송합니다 🙇🏻‍♂️

1. 폼 정보 전부 입력 후 인증 요청 클릭
2. 인증번호 확인에 123458 쓰고 인증하기 클릭
3. 인증 실패 후 회원가입에 탭으로 넘어가서 엔터 누르기
4. 가입 성공

https://github.com/user-attachments/assets/1f6edf50-15a2-4b24-a8c6-da16c129b6b0

아마 completeVerification의 위치가 바뀌면 자연스럽게 해결 될 것 같습니다 👍👍

kimyou1102 replied:
이 오류도 submit핸들러에서 `verificationCodeError` 조건을 추가해서 해결했습니다! 꼼꼼한 리뷰 감사합니다~~

kimyou1102 replied:
verificationCodeError조건이 아닌 인증코드 입력의 예외검증 값인 `verificationCodeValidated` 를 조건에 추가해서 해결했습니다!

[fix: 인증 성공 시에만 verified 상태로 전환되도록 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/afb839fa7cd1526f6a62c12e3f1071e2f9f1d180)

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -132,19 +158,50 @@ function SignupForm() {
+      });
+    } finally {
+      completeVerification();
+    }
```
p4: 혹시 completeVerification이 try문 내부의 인증 성공 시에만 동작하는 건 어떻게 생각하시나요? 🤔

현재는 finally에 의해 인증이 실패해도 verificationStep이 verified가 됩니다. 

인증이 실패했는데 인증이 완료되었다는 상태가 되는 게 어색해서 말해봤습니다.

또한 현재 코드로는 handleSubmit의 조건을 통과할 수 있었습니다!

```tsx
if (verificationStep !== 'verified') {  // ← 제출 가능 
  return;
}
  ```

kimyou1102 replied:
말씀해주신 것처럼 `completeVerification`은 성공 시에만 호출하는 것이 맞다고 생각합니다.
기존에는 `verified` 상태와 에러 메시지 노출 로직이 결합되어 있어, 인증 실패 시에도 에러 메시지를 보여주기 위해 성공/실패와 무관하게 상태를 변경했었습니다.

이번 수정에서는 성공 시에만 `completeVerification`을 호출하고, 에러 메시지 로직을 `verified` 상태와 분리해 구조를 정리하겠습니다. 덕분에 상태 전이와 로직을 한 번 더 점검할 수 있었습니다. 감사합니다!

kimyou1102 replied:
[fix: 인증 성공 시에만 verified 상태로 전환되도록 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/afb839fa7cd1526f6a62c12e3f1071e2f9f1d180)

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -153,7 +210,7 @@ function SignupForm() {
     }
 
-    return errorMessage === '' ? submitVerificationErrorMessage : errorMessage;
+    return errorMessage;
```
p5: getDisplayedVerificationErrorMessage가 항상 errorMessage를 반환해서 내부에 있는 조건이 없어져도 될 거 같아요 😊

kimyou1102 replied:
헉맞네요 로직을 수정하면서 필요없어져서 함수자체가 필요가 없어지네요 불필요한 getDisplayedVerificationErrorMessage를 제거하게끔 수정하겠습니다!

kimyou1102 replied:
[refactor: 의미 없는 getDisplayedVerificationErrorMessage 제거 및 getFinalVerificationCodeErrorMessage 직접 사용으로 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/feb07430e9153d4b157c59524d488255941b5338)

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -153,7 +210,7 @@ function SignupForm() {
+    return errorMessage;
   };
 
   const validateForm = () => {
```
p5: 혹시 validateForm에는 verificationStep이 verified인지 확인할 필요는 없는지 궁금합니다 🤔

verificationCodeError가 false라면 verificationStep이 verified가 보장되어서 혹시 없는 걸까요??

kimyou1102 replied:
인증요청/인증하기 버튼 클릭을 누락했을 때 어디에서 문제가 발생했는지 바로 알려주기 위해, 제출 버튼은 비활성화하지 않고 제출 이벤트만 가드하도록 구현했습니다!
이렇게 하면 제출 시점에 잘못된 입력 필드에 즉시 에러 메시지가 노출되어 UX가 더 명확하다고 판단했습니다. 로건은 이런 방식에 대해서는 어떠신가요??

dlsxjzld replied:
잘못된 입력 필드가 한 눈에 보여서 현재는 상관 없을 것 같아요 👍
만약 입력 필드가 늘어나서 에러 메시지가 노출되는데 화면에서 보이지 않는다면 그 때는 제출 버튼을 비활성화해도 좋을 것 같아요 👍

저도 한 번 더 생각해볼 수 있어서 좋았습니다! 감사합니다 ☺️

dlsxjzld commented on frontend/src/pages/signup/components/SignupForm/SignupForm.tsx:
```diff
@@ -205,21 +277,13 @@ function SignupForm() {
 
   const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
```
p2: msw로 테스트를 해봤는데 이런 버그가 있었습니다. 😊


https://github.com/user-attachments/assets/59c5f1f6-cdf9-491d-a833-8927dc73d877



1. 123456 입력 → 인증 성공
   - verificationCodeError = false
   - verificationStep = 'verified'

2. 인증번호를 12345로 변경
   - verificationCodeValidated = false (6자 미만)
   - verificationCodeError = false (아직 서버 검증 안 함)
   - verificationStep = 'verified' (그대로 유지)

3. validateForm() 체크:
   - verificationCodeValidated && !verificationCodeError
   - false && true = false
   - → 버튼 비활성화 ✅ (정상)

4. 엔터키로 제출:
   - handleSubmit에서 verificationStep === 'verified' 체크
   - 'verified' === 'verified' = true
   - → 제출 진행! ❌ (버그)

handleSubmit과 validateForm이 다른 기준으로 검증이 되어 발생하는 것 같습니다! 🤔
handleSubmit에 verificationCodeValidate를 추가하거나 아니면 validateForm을 재사용하면 해결될 것 같습니다!

kimyou1102 replied:
엔터키는 생각을 못했는데 리뷰 덕분에 찾아냈네요 꼼꼼한 리뷰 감사합니다!!👍👍

kimyou1102 replied:
[fix: 인증 성공 시에만 verified 상태로 전환되도록 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/1134/commits/afb839fa7cd1526f6a62c12e3f1071e2f9f1d180)

dlsxjzld reviewed (commented):
리팩터링 및 테스트 코드 꼼꼼하게 작성하느라 고생 많으셨습니다!!! 👍👍

몇가지 궁금증과 발견한 버그에 대해서 코멘트 남겨놨습니다! 😊

kimyou1102 committed:
refactor: dotenv config 경로를 절대에서 상대 경로로 변경 #733

kimyou1102 committed:
refactor: 불필요한 주석 제거 #733

kimyou1102 committed:
fix: handleSubmit에 verificationCodeError 조건 추가하여
인증 실패 시 폼 제출 방지 #733

kimyou1102 committed:
refactor: 의미 없는 getDisplayedVerificationErrorMessage 제거 및
getFinalVerificationCodeErrorMessage 직접 사용으로 변경 #733

kimyou1102 committed:
fix: 인증 성공 시에만 verified 상태로 전환되도록 수정 #733

- 인증 성공 케이스에서만 인증 단계를 verified로 변경하도록 로직 수정
- submit 시 입력 검증 실패 시 엔터 제출이 발생하지 않도록 가드 조건을 개선

kimyou1102 committed:
refactor: 아이디 중복확인 상태를 duplicateChecked상태로 판단하도록 변경 #733

- 중복확인 여부를 현재 userId 기준의 duplicateChecked 상태로 판단하도록 변경
- 아이디 입력 변경 시 기존 중복확인 결과가 자동으로 무효화되도록 개선

kimyou1102 committed:
refactor: MSW 핸들러 URL에서 환경변수를 읽기 위해 env 로딩을 vitest.config.ts로 이동 #733

kimyou1102 committed:
refactor: MSW 핸들러 URL 와일드카드 제거 #733

kimyou1102 commented:
@dlsxjzld 리뷰 반영 완료했습니다! 반영사항 확인하시고 추가로 수정사항 있다면 말씀해주세요~~

dlsxjzld commented:
> @dlsxjzld 리뷰 반영 완료했습니다! 반영사항 확인하시고 추가로 수정사항 있다면 말씀해주세요~~

확인했습니다! 👍👍
